src/leap/mail/mail.py

```diff
@@ -0,0 +1,697 @@
+
+        d = self.all_uid_iter()
+        d.addCallback(del_all_uid)
+        return d
+
+    def _update_flags_or_tags(self, old, new, mode):
+        if mode == Flagsmode.APPEND:
+            final = list((set(tuple(old) + new)))
+        elif mode == Flagsmode.REMOVE:
+            final = list(set(old).difference(set(new)))
+        elif mode == Flagsmode.SET:
+            final = new
+        return final
+
+    def udpate_flags(self, msg, flags, mode):
+        """
+        Update flags for a given message.
+        """
+        wrapper = msg.get_wrapper()
+        current = wrapper.fdoc.flags
+        newflags = self._update_flags_or_tags(current, flags, mode)
+        wrapper.fdoc.flags = newflags
+
+        wrapper.fdoc.seen = MessageFlags.SEEN_FLAG in newflags
+        wrapper.fdoc.deleted = MessageFlags.DELETED_FLAG in newflags
+
+        return self.adaptor.update_msg(self.store, msg)
+
+    def update_tags(self, msg, tags, mode):
+        """
+        Update tags for a given message.
+        """
+        wrapper = msg.get_wrapper()
+        current = wrapper.fdoc.tags
+        newtags = self._update_flags_or_tags(current, tags, mode)
+        wrapper.fdoc.tags = newtags
+        return self.adaptor.update_msg(self.store, msg)
+
+
+class Account(object):
+    """
+    Account is the top level abstraction to access collections of messages
+    associated with a LEAP Mail Account.
+
+    It primarily handles creation and access of Mailboxes, which will be the
+    basic collection handled by traditional MUAs, but it can also handle other
+    types of Collections (tag based, for instance).
+
+    leap.mail.imap.SoledadBackedAccount partially proxies methods in this
+    class.
+    """
+
+    # Adaptor is passed to the returned MessageCollections, so if you want to
+    # use a different adaptor this is the place to change it, by subclassing
+    # the Account class.
+
+    adaptor_class = SoledadMailAdaptor
+    store = None
+
+    def __init__(self, store, ready_cb=None):
+        self.store = store
+        self.adaptor = self.adaptor_class()
+        self.mbox_indexer = MailboxIndexer(self.store)
+
+        self.deferred_initialization = defer.Deferred()
+        self._ready_cb = ready_cb
+
+        self._init_d = self._initialize_storage()
+
+    def _initialize_storage(self):
+
+        def add_mailbox_if_none(mboxes):
+            if not mboxes:
+                return self.add_mailbox(INBOX_NAME)
+
+        def finish_initialization(result):
+            self.deferred_initialization.callback(None)
+            if self._ready_cb is not None:
+                self._ready_cb()
+
+        d = self.adaptor.initialize_store(self.store)
+        d.addCallback(lambda _: self.list_all_mailbox_names())
+        d.addCallback(add_mailbox_if_none)
+        d.addCallback(finish_initialization)
+        return d
+
+    def callWhenReady(self, cb, *args, **kw):
+        self.deferred_initialization.addCallback(cb, *args, **kw)
```

Comment:
As mentioned above, I'm not sure if I would prefer to do something like:
```
self.deferred_initialization.addCallback(lambda _: cb(*args, **kw))
```
If we keep it like that I think we should document that `cb` will get a meaningless first parameter.